### PR TITLE
Another approach to fixing deployNodes task and network parameters generation

### PR DESCRIPTION
**Pull request by kasiastreich**
As discussed, I use class loading approach in `deployNodes` gradle task just after the `NodeInfos` generation. Unfortunately, I had some crypto packages loading problems that need to be solved. As a quick fix I disabled checking of the signature for cordformation in devMode.

### CODE REVIEW

File: gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt [hunk 1]
```diff
@@ -28,8 +27,10 @@ open class Cordform : DefaultTask() {
      */
     @Suppress("MemberVisibilityCanPrivate")
     var definitionClass: String? = null
+    var parametersClass: String? = "net.corda.node.internal.networkParametersGenerator.TestNetworkParametersGenerator"
```
Comment (Clintonio) [inline gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt]: This should be a private val, not public var.

File: gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt [hunk 2]
```diff
@@ -142,14 +157,39 @@ open class Cordform : DefaultTask() {
         }
     }
 
+    private fun generateAndInstallNetworkParameters() {
+        project.logger.info("Generating network parameters")
+        gatherNotaries()
```
Comment (Clintonio) [inline gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt]: This should return the value and use it locally instead of storing in a class variable, because it isn't used outside of this function.

Comment (Clintonio) [reply gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt]: (Function and callees)

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/ParametersUtilities.kt [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package net.corda.testing.common.internal
+package net.corda.nodeapi.internal
```
Comment (Clintonio) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/ParametersUtilities.kt]: Do these need to be in nodeapi?

File: node/src/main/kotlin/net/corda/node/services/network/NodeInfoWatcher.kt [hunk 4]
```diff
@@ -91,20 +94,55 @@ class NodeInfoWatcher(private val nodePath: Path,
...
 
     /**
+     * Loads all NodeInfo files stored in node's base directory and returns the [Party]s.
+     * Scans main directory and [CordformNode.NODE_INFO_DIRECTORY].
+     * Signatures are not checked before returning a value. The latest value stored for a given name is returned.
+     *
+     * @return a list of [Party]s
+     */
+    fun loadAndGatherNotaryIdentities(notaries: Map<CordaX500Name, Boolean>): List<NotaryInfo> {
+        val nodeInfos = loadFromDirectory(false)
```
Comment (Clintonio) [inline node/src/main/kotlin/net/corda/node/services/network/NodeInfoWatcher.kt]: 104-115 could be a separate, private, function called something like "getAllNodeInfos".

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/ParametersUtilities.kt (hunk 3 above)
Comment (kasiastreich) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/ParametersUtilities.kt]: I needed to have an access to it from node, any better place to have it?

File: node/src/main/kotlin/net/corda/node/services/network/NodeInfoWatcher.kt [hunk 5]
```diff
@@ -91,20 +94,60 @@ class NodeInfoWatcher(private val nodePath: Path,
...
+                notaryInfos.add(NotaryInfo(notaryId, notaries[info.legalIdentities[0].name]!!))
+            }
+        }
+        return notaryInfos.toList()
+    }
+
+    private fun getAllNodeInfos(): Map<CordaX500Name, NodeInfo> {
+        val nodeInfos = loadFromDirectory(false)
+        // NodeInfos are currently stored in 2 places: in [CordformNode.NODE_INFO_DIRECTORY] and in baseDirectory of the node.
+        val myFiles = nodePath.list {it.filter { it.toString().contains("nodeInfo-") }.toList() }
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/network/NodeInfoWatcher.kt]: `in` instead of `contains`

Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/network/NodeInfoWatcher.kt]: Space after `{`

File: node/src/main/kotlin/net/corda/node/services/network/NodeInfoWatcher.kt [hunk 6]
```diff
@@ -91,20 +94,60 @@ class NodeInfoWatcher(private val nodePath: Path,
...
+        val myNodeInfos = myFiles.mapNotNull { processFile(it, false) }
+        val infosMap = mutableMapOf<CordaX500Name, NodeInfo>()
+        // Running deployNodes more than once produces new NodeInfos. We need to load the latest NodeInfos based on serial field.
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/network/NodeInfoWatcher.kt]: This is not a feature but a bug. I thought this was fixed? If it's not can you add a TODO in the relevant bit of the code. Assume there's only one nodeInfo file in the baseDirectory.

File: node/src/main/kotlin/net/corda/node/services/network/NodeInfoWatcher.kt [hunk 7]
```diff
@@ -91,20 +94,60 @@ class NodeInfoWatcher(private val nodePath: Path,
...
+    fun loadAndGatherNotaryIdentities(notaries: Map<CordaX500Name, Boolean>): List<NotaryInfo> {
+        val infosMap = getAllNodeInfos()
+        // Now get the notary identities based on names passed from Cordform configs. There is one problem, for distributed notaries
+        // Cordfom specifies in config only node's main name, the notary identity isn't passed there. It's read from keystore on
+        // node startup, so we have to look it up from node info as a second identity, which is ugly.
+        val notaryInfos = mutableSetOf<NotaryInfo>()
+        for (info in infosMap.values) {
+            // Here the ugliness happens.
+            // TODO Change Cordform definition so it specifies distributed notary identity.
+            if (info.legalIdentities[0].name in notaries.keys) {
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/network/NodeInfoWatcher.kt]: You can use a series of `map`, `filter`, `duplicate`, etc on `infosMap.value` to do the same thing. It will probably look cleaner

File: samples/simm-valuation-demo/build.gradle [hunk 8]
```diff
@@ -36,6 +36,7 @@ dependencies {
     cordaCompile project(path: ":node:capsule", configuration: 'runtimeArtifacts')
     cordaCompile project(path: ":webserver:webcapsule", configuration: 'runtimeArtifacts')
     cordaCompile project(':core')
+    cordaCompile project(':node')
```
Comment (shamsasari) [inline samples/simm-valuation-demo/build.gradle]: The demos are not supposed to have a dependency to the node. Why is this needed?

File: node/src/main/kotlin/net/corda/node/services/network/NodeInfoWatcher.kt (hunk 6 above)
Comment (kasiastreich) [reply node/src/main/kotlin/net/corda/node/services/network/NodeInfoWatcher.kt]: When i was working on that it wasn't fixed. Need to check if there is a commit that addresses this issue.

File: gradle-plugins/cordform-common/src/main/java/net/corda/cordform/NetworkParametersGenerator.java [hunk 9]
```diff
@@ -0,0 +1,9 @@
+package net.corda.cordform;
+
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Map;
+
+public interface NetworkParametersGenerator {
+    void run(Map<String, Boolean> notaryMap, List<Path> nodesDirs);
```
Comment (shamsasari) [inline gradle-plugins/cordform-common/src/main/java/net/corda/cordform/NetworkParametersGenerator.java]: Document the parameters

File: gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Node.kt [hunk 10]
```diff
@@ -124,14 +123,11 @@ class Node(private val project: Project) : CordformNode() {
...
-            if (o.size > 0) {
-                o.first().first.value.toString()
-            } else {
-                name
-            }
-        } catch(_ : IllegalArgumentException) {
-            // Can't parse as an X500 name, use the full string
+            // Using sun.security.x509.X500Name because importing BouncyCastle provider in Cordformation causes problems
+            // with loading our custom X509EdDSAEngine.
+            X500Name(name).organization
```
Comment (shamsasari) [inline gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Node.kt]: Why not just write a regex to extract out the `O=` bit of the string? Seems better than to import BC or use a private internal API.

File: gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordformation.kt [hunk 11]
```diff
@@ -31,5 +31,11 @@ class Cordformation : Plugin<Project> {
 
     override fun apply(project: Project) {
         Utils.createCompileConfiguration("cordapp", project)
+        // Add runtime dependency to node except for the corda project itself to avoid a circular dependency of Corda with itself.
```
Comment (shamsasari) [inline gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordformation.kt]: Perhaps add a comment on why this needs to be done

File: node/src/main/kotlin/net/corda/node/internal/networkParametersGenerator/TestNetworkParametersGenerator.kt [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        return infos.filter { it.legalIdentities[0].name in notaries.keys }.map {
+            NotaryInfo(it.notaryIdentity(), notaries[it.legalIdentities[0].name]!!)
+        }.distinct()
+    }
+
+    private fun NodeInfo.notaryIdentity() = if (legalIdentities.size == 2) legalIdentities[1] else legalIdentities[0]
+
+    private fun initialiseSerialization() {
+        val context = if (java.lang.Boolean.getBoolean("net.corda.testing.amqp.enable"))
+            AMQP_P2P_CONTEXT
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/networkParametersGenerator/TestNetworkParametersGenerator.kt]: I think this if-else is small enough to all go on the line above

File: node/src/main/kotlin/net/corda/node/services/network/NodeInfoWatcher.kt [hunk 13]
```diff
@@ -90,6 +93,29 @@ class NodeInfoWatcher(private val nodePath: Path,
...
     fun saveToFile(signedNodeInfo: SignedData<NodeInfo>) = Companion.saveToFile(nodePath, signedNodeInfo)
 
+    /**
+     * Loads latest NodeInfo files stored in node's base directory.
+     * Scans main directory and [CordformNode.NODE_INFO_DIRECTORY].
+     * Signatures are checked before returning a value. The latest value stored for a given name is returned.
+     *
+     * @return list of latest [NodeInfo]s
+     */
+    fun getAllNodeInfos(): List<NodeInfo> {
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/network/NodeInfoWatcher.kt]: I still think this should be in `TestNetworkParametersGenerator`. Just make `loadFromDirectory` public.

File: gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Node.kt [hunk 14]
```diff
@@ -121,12 +119,11 @@ class Node(private val project: Project) : CordformNode() {
...
         }
-
         val dirName = try {
-            // Using sun.security.x509.X500Name because importing BouncyCastle provider in Cordformation causes problems
+            // Parsing O= part directly because importing BouncyCastle provider in Cordformation causes problems
             // with loading our custom X509EdDSAEngine.
-            X500Name(name).organization
-        } catch(_ : IOException) {
+            name.trim().split(",").filter { it.startsWith("O=") }.first().substringAfter("=")
+        } catch(_ : NoSuchElementException) {
```
Comment (shamsasari) [inline gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Node.kt]: use `firstOrNull {  }`to avoid catching this exception

File: gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Node.kt [hunk 15]
```diff
@@ -119,14 +119,10 @@ class Node(private val project: Project) : CordformNode() {
...
-            // Parsing O= part directly because importing BouncyCastle provider in Cordformation causes problems
-            // with loading our custom X509EdDSAEngine.
-            name.trim().split(",").filter { it.startsWith("O=") }.first().substringAfter("=")
-        } catch(_ : NoSuchElementException) {
-            // Can't parse as an X500 name or no organisation part, use the full string
-            name
-        }
+        // Parsing O= part directly because importing BouncyCastle provider in Cordformation causes problems
+        // with loading our custom X509EdDSAEngine.
+        val organizationName = name.trim().split(",").filter { it.startsWith("O=") }.firstOrNull()?.substringAfter("=")
```
Comment (shamsasari) [inline gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Node.kt]: Can you merge the filter and firstOrNull with just firstOrNull {}

File: constants.properties [hunk 16]
```diff
@@ -1,4 +1,4 @@
-gradlePluginsVersion=2.0.8
+gradlePluginsVersion=2.0.9
```
Comment (shamsasari) [inline constants.properties]: Append -NETWORKMAP to avoid any conflict with master

File: gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt [hunk 17]
```diff
@@ -28,6 +27,7 @@ open class Cordform : DefaultTask() {
      */
     @Suppress("MemberVisibilityCanPrivate")
     var definitionClass: String? = null
+    private val parametersClass: String = "net.corda.node.internal.networkParametersGenerator.TestNetworkParametersGenerator"
```
Comment (shamsasari) [inline gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt]: networkParametersGenClass

File: node/src/main/kotlin/net/corda/node/internal/networkParametersGenerator/TestNetworkParametersGenerator.kt [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+import net.corda.nodeapi.internal.serialization.*
+import net.corda.nodeapi.internal.serialization.amqp.AMQPServerSerializationScheme
+import java.nio.file.Path
+import java.time.Instant
+import kotlin.streams.toList
+
+// This class is used by deployNodes task to generate NetworkParameters in [Cordformation].
+class TestNetworkParametersGenerator : NetworkParametersGenerator {
+    override fun run(notaryMap: Map<String, Boolean>, nodesDirs: List<Path>) {
+        initialiseSerialization()
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/networkParametersGenerator/TestNetworkParametersGenerator.kt]: Shouldn't we also be turning off the serialisation context at the end of this method?

File: gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Node.kt [hunk 19]
```diff
@@ -121,7 +121,7 @@ class Node(private val project: Project) : CordformNode() {
         }
         // Parsing O= part directly because importing BouncyCastle provider in Cordformation causes problems
         // with loading our custom X509EdDSAEngine.
-        val organizationName = name.trim().split(",").filter { it.startsWith("O=") }.firstOrNull()?.substringAfter("=")
+        val organizationName = name.trim().split(",").firstOrNull { it.startsWith("O=") } ?.substringAfter("=")
```
Comment (shamsasari) [inline gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Node.kt]: remove space before ?.

File: node/src/main/kotlin/net/corda/node/internal/networkParametersGenerator/TestNetworkParametersGenerator.kt [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+     * Loads latest NodeInfo files stored in node's base directory.
+     * Scans main directory and [CordformNode.NODE_INFO_DIRECTORY].
+     * Signatures are checked before returning a value. The latest value stored for a given name is returned.
+     *
+     * @return list of latest [NodeInfo]s
+     */
+    private fun getAllNodeInfos(baseDirectory: Path): List<NodeInfo> {
+        val infoWatcher = NodeInfoWatcher(baseDirectory)
+        val nodeInfos = infoWatcher.loadFromDirectory()
+        // NodeInfos are currently stored in 2 places: in [CordformNode.NODE_INFO_DIRECTORY] and in baseDirectory of the node.
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/networkParametersGenerator/TestNetworkParametersGenerator.kt]: Why do you even need to scan NODE_INFO_DIRECTORY when you can look at the baseDirectory? And if we just need to look at baseDirectory do we even need to use the NodeInfoWatcher? `processFile` is straightforward enough to re-implement.

File: node/src/main/kotlin/net/corda/node/internal/networkParametersGenerator/TestNetworkParametersGenerator.kt [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+        val myFiles = baseDirectory.list { it.filter { "nodeInfo-" in it.toString() }.toList() }
+        val myNodeInfos = myFiles.mapNotNull { infoWatcher.processFile(it) }
+        val infosMap = mutableMapOf<CordaX500Name, NodeInfo>()
+        // Running deployNodes more than once produces new NodeInfos. We need to load the latest NodeInfos based on serial field.
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/networkParametersGenerator/TestNetworkParametersGenerator.kt]: Let's not work around this bug as if it's a feature. Just take the first node info file.

File: node/src/main/kotlin/net/corda/node/internal/networkParametersGenerator/TestNetworkParametersGenerator.kt [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+import net.corda.nodeapi.internal.NetworkParametersCopier
+import net.corda.nodeapi.internal.serialization.*
+import net.corda.nodeapi.internal.serialization.amqp.AMQPServerSerializationScheme
+import java.nio.file.Path
+import java.time.Instant
+import kotlin.streams.toList
+
+// This class is used by deployNodes task to generate NetworkParameters in [Cordformation].
+class TestNetworkParametersGenerator : NetworkParametersGenerator {
+    override fun run(notaryMap: Map<String, Boolean>, nodesDirs: List<Path>) {
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/networkParametersGenerator/TestNetworkParametersGenerator.kt]: Since we have all the node directories, can we move more of the processing into this method, namely removing the need for `notaryMap` param. Parse each node.conf file and extract out the notary config.

File: node/src/main/kotlin/net/corda/node/internal/networkParametersGenerator/TestNetworkParametersGenerator.kt (hunk 20 above)
Comment (kasiastreich) [reply node/src/main/kotlin/net/corda/node/internal/networkParametersGenerator/TestNetworkParametersGenerator.kt]: Because I was using only one node to generate the parameters based on all information it had. I can change it to scan all nodes, so we need only one nodeInfo from each `baseDirectory`.

Comment (shamsasari) [reply node/src/main/kotlin/net/corda/node/internal/networkParametersGenerator/TestNetworkParametersGenerator.kt]: That makes more sense

File: node/src/main/kotlin/net/corda/node/internal/networkParametersGenerator/TestNetworkParametersGenerator.kt (hunk 18 above)
Comment (kasiastreich) [reply node/src/main/kotlin/net/corda/node/internal/networkParametersGenerator/TestNetworkParametersGenerator.kt]: I am not sure how?

Comment (shamsasari) [reply node/src/main/kotlin/net/corda/node/internal/networkParametersGenerator/TestNetworkParametersGenerator.kt]: Have a look at the companion object of `NodeProcess.Factory`

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/TestNetworkParametersGenerator.kt [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+        val context = if (java.lang.Boolean.getBoolean("net.corda.testing.amqp.enable")) AMQP_P2P_CONTEXT else KRYO_P2P_CONTEXT
+        _contextSerializationEnv.set(SerializationEnvironmentImpl(
+                SerializationFactoryImpl().apply {
+                    registerScheme(KryoParametersSerializationScheme)
+                    registerScheme(AMQPServerSerializationScheme())
+                },
+                context))
+    }
+
+    private object KryoParametersSerializationScheme : AbstractKryoSerializationScheme() {
```
Comment (shamsasari) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/TestNetworkParametersGenerator.kt]: A comment perhaps why we need this. You were using the stock scheme before.

File: gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt [hunk 24]
```diff
@@ -113,6 +113,19 @@ open class Cordform : DefaultTask() {
...
+    /**
+     * The parametersGenerator needn't be compiled until just before our build method, so we load it manually via sourceSets.main.runtimeClasspath.
+     */
+    private fun loadParametersGenerator(): NetworkParametersGenerator {
+        val plugin = project.convention.getPlugin(JavaPluginConvention::class.java)
+        val classpath = plugin.sourceSets.getByName(MAIN_SOURCE_SET_NAME).runtimeClasspath
+        val urls = classpath.files.map { it.toURI().toURL() }.toTypedArray()
+        return URLClassLoader(urls, NetworkParametersGenerator::class.java.classLoader)
+                .loadClass(networkParametersGenClass)
+                .asSubclass(NetworkParametersGenerator::class.java)
```
Comment (shamsasari) [inline gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt]: Because we already know the class to instantiate, do you think we could get away with removing the `NetworkParametersGenerator` interface entirely and just call the `run` method using reflection? `NetworkParametersGenerator` is giving us compile-time safety but nothing more than that. WDYT?

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/TestNetworkParametersGenerator.kt (hunk 23 above)
Comment (kasiastreich) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/TestNetworkParametersGenerator.kt]: It's because the previous one isn't accessible from `node-api`.

Comment (shamsasari) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/TestNetworkParametersGenerator.kt]: I mean in a comment in the code

File: constants.properties [hunk 25]
```diff
@@ -1,4 +1,4 @@
-gradlePluginsVersion=2.0.9
+gradlePluginsVersion=2.0.9-NETWORKMAP
```
Comment (Clintonio) [inline constants.properties]: Please change to 3.0.0. Also why the -NETWORKMAP?

File: gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt (hunk 24 above)
Comment (Clintonio) [reply gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt]: I'm fine keeping it in because in the case this goes wrong the exceptions produced will be much more helpful.

File: gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Node.kt [hunk 26]
```diff
@@ -250,34 +239,17 @@ class Node(private val project: Project) : CordformNode() {
...
-    /**
-     * Find the corda JAR amongst the dependencies
+     * Find the given JAR amongst the dependencies
+     * @param jarName JAR name without the version part, for example for corda-2.0-SNAPSHOT.jar provide only "corda" as jarName
      *
-     * @return A file representing the Corda webserver JAR
+     * @return A file representing found JAR
      */
-    private fun verifyAndGetWebserverJar(): File {
+    private fun verifyAndGetJar(jarName: String): File {
```
Comment (Clintonio) [inline gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Node.kt]: Please rename to `verifyAndGetRuntimeJar`

File: constants.properties (hunk 25 above)
Comment (shamsasari) [reply constants.properties]: Oops!

Comment (kasiastreich) [reply constants.properties]: NETWORKMAP was requested by Shams.

Comment (shamsasari) [reply constants.properties]: (Btw, ignore my "Oops" I thought I was looking at my own PR!)

W.r.t. to the -NETWORKMAP here, we need it to so that we don't clash with any gradle plugins on master. This will be removed once we merge this feature branch into master.

File: constants.properties [hunk 27]
```diff
@@ -1,4 +1,4 @@
-gradlePluginsVersion=2.0.9-NETWORKMAP
+gradlePluginsVersion=3.0.0-NETWORKMAP
```
Comment (Clintonio) [inline constants.properties]: What is the purpose of -NETWORKMAP here?

Comment (shamsasari) [reply constants.properties]: https://github.com/corda/corda/pull/2066#discussion_r153797980
